### PR TITLE
upgrade css/js optimizer to ezpEvent logic

### PR DESCRIPTION
**Pull request by jdespatis**
The Pull is speaking: "Don't merge me, I'm buggy!"...

It's just a pull to start a discussion, because I have 2 problems with current ezp logic

First and foremost, there's an event response/output to trigger some callbacks whenever the html is pushed to the client, thanks to ezpEvent

This feature is really nice, and even so nice that it may be interesting to put those ezpEvent everywhere!

And why not in ezjscore, whenever it wants to push css or js to the client (just like html indeed)?

=> So I would say it would be interesting to create 2 new events: response/outputcss and response/outputjs (or even a more cosmetic response/output/css, response/output/js), just like the current response/output (which could be renamed into response/output/html)

Of course, feel free to tell me if you find first this idea stupid, but even if it is, it's interesting to read below

You can quickly look at the pull to understand how it can easily be done

BUT it's buggy, in fact it's currently impossible to call a filter with more than one argument, which is the value that the filter may alter

1/ Here I find this logic to be too restrictive, it would be nice to call a filter with a value, and maybe another param, for example the context (in this pull, the second param needed would be the packLevel), and in the long run as many params as possible
=> so why not add some logic of func_get_args() in the ezpEvent::filter() function ? the ezpEvent::notify() already supports an array as param where it's possible to push many arguments, but it doesn't alter the value, which is what I need here

2/ and second, imagine now this pull is correct (it would if ezpEvent::filter() could support many args). I have still a problem...
Ok, the code would work, it could minify css and js
But imagine a developer wants then to create another extension that bundles custom minifiers for css and js, and he doesn't want the default js/css minifier of ezjscore to be triggered at all

=> how can he/she deletes those lines? :

`````` Listeners[]=response/outputcss@ezjscCssOptimizer::optimize
Listeners[]=response/outputjs@ezjscJavascriptOptimizer::optimize```

and put his/her classes ?
Of course, it's impossible to do Listeners[]= to empty (because it would empty all callbacks for all filters)
The developer would only want to empty the callbacks triggered by response/outputcss, in order to put his/her callbacks

If the syntax was: 
```Listeners[response/outputcss]=ezjscCssOptimizer::optimize```

Then the developer could empty Listeners[response/outputcss] and afterwards put his/her classes. But of course this syntax is too limitative as it's impossible to put more than one callback to an event. It's just to explain the problem I have to tackle with

=> You see a simple solution ?

(the first I can see if for the developer to have his/her extension loaded first, and detach asap all other callbacks for the response/outputcss for example, but not totally sure it would work, because this detach may occur too late, after the foreach of ezpEvent::filter())
Whatever, this trick is more complex then just performing the common Setting[]= to empty the array and feed it from the start

Any idea on this logic for settings Listeners in site.ini ?
``````

### CODE REVIEW

**Comment by andrerom** [comment]
> Any idea on this logic for settings Listeners in site.ini ?

Use a unique array key in the setting os it can be overriden:

```
Listeners[css_optimizer]=response/outputcss@ezjscCssOptimizer::optimize
```

It is still in a extension though, so you override these settings form siteaccess for instance.

**Comment by jdespatis** [comment]
ok @andrerom nice!

and what about the handle of many args in the ezpEvent::filter() (with some logic with func_get_args()), do you think it would be nice ?

is it possible to have a filter like response/output/css or only response/outputcss ?

**Comment by andrerom** [comment]
"nice" is maybe not the word I would use, but it might be useful.
Atm you can potentially do it by using ezpEvent::notify() with string as a reference.

**Comment by jdespatis** [comment]
ok patch updated

it assumes that some get_func_args logic exists in ezpEvent::filter(), which is not yet the case, I'll do a pull for that

also missing a file in 4.7 to show the 2 new filters

And maybe if @andrerom you're ok with this pull,
then it might be interesting to rename response/output to response/output/html and response/preoutput to response/preoutput/html ? (to keep an homogeneity)

**Comment by andrerom** [comment]
> might be interesting to rename response/output to response/output/html and response/preoutput to response/preoutput/html ? (to keep an homogeneity)

Not really as that would be a break (even though the feature is marked as private).

File: classes/ezjscpacker.php [hunk 1]
```diff
@@ -458,13 +458,7 @@ static function packFiles( $fileArray, $subPath = '', $fileExtension = '.js', $p
...
 
         // Pack all files to save bandwidth
-        if ( $data['pack_level'] > 1 )
-        {
-            foreach( $ezjscINI->variable( 'eZJSCore', $isCSS ? 'CssOptimizer' : 'JavaScriptOptimizer' ) as $optimizer )
-            {
-                $content = call_user_func( array( $optimizer, 'optimize' ), $content, $data['pack_level'] );
-            }
-        }
+        $content = ezpEvent::getInstance()->filter( $isCSS ? 'response/output/css' : 'response/output/javascript', $content, $data['pack_level'] );
```
Comment (andrerom) [inline classes/ezjscpacker.php]: Should be parentheses around

```
$isCSS ? 'response/output/css' : 'response/output/javascript'
```

**Comment by jdespatis** [comment]
Yes it would break, but this feature is very recent, not sure a lot of people already uses it, so it's a chance to change it now no?

or maybe a better way would be to deprecate the response/output and index.php (and so on) triggers both response/output and response/output/html until 4.7

and at 4.7 response/output is vanished

Do you think it would help?

**Comment by jdespatis** [comment]
@andrerom like this ?
I would not add such parentheses but as you like :)

or maybe parenthesis around $isCSS ?

**Comment by andrerom** [comment]
> Yes it would break, but this feature is very recent, not sure a lot of people already uses it, so it's a chance to change it now no?

ezformtoken uses response/output, it was bundled with 4.6.
And furthermore the output might not be html.

**Comment by andrerom** [comment]
If you ask me, you should rather rename the filter name used here, it is not response output, it is usually written to disk before being requested by client..

**Comment by andrerom** [comment]
> I would not add such parentheses but as you like :)

Yes, makes the code more readable, could also have a new line after each comma.

**Comment by jdespatis** [comment]
ok thanks for those info

No problem to change the name of those filters, some proposals ? :)

The fact is I'm best to find a name and change it afterwards...

**Comment by jdespatis** [comment]
store/css 
store/javascript ?

to find a good name, it would be nice to imagine all other operations possible to find a generic one that would be used by those filters and by other

If you have a good idea, you're welcome :)

**Comment by jdespatis** [comment]
or maybe just optimize/css and optimize/javascript ?

**Comment by andrerom** [comment]
Think I like optimize/*.

@bdunogier What do you think?

**Comment by jdespatis** [comment]
btw @andrerom , for this commit, a file is needed somewhere to explain those 2 new filters
=> where would you put it ? in ezpublish/doc/features/4.7/event.txt or in something like ezjscore/doc/features/1.x/event.txt ?

**Comment by andrerom** [comment]
first maybe

**Comment by jdespatis** [comment]
Ok, I change the filter name

@bdunogier what do you think of optimize/css and optimize/javascript for those 2 new filters ?

@andrerom tell me if you think this commit is complete or not (it needs another pull request but on another git project)

**Comment by jdespatis** [comment]
After some reflexion, I don't think the trick to add a key in Listeners[], to let the developers delete the optimizer from ezjscore, is a good idea, see below why:

1/ first, I'm not sure this logic could be well understood by a developer, it might seem weird to use such keywords logic, in ezjscore, and in fact only in ezjscore...
2/ and second, which is indeed the main reason, it doesn't help so much the developer. Imagine someone has 2 custom extensions, each of them wants to register a new tool to optimize js for example. It's complex to do it indeed, because the developer should make sure the custom extensions are loaded in a specific order, in order for one extension (and only one) to delete the optimizer of ezjscore by doing "Listeners[javascript_optimizer]=" in site.ini.append.php

Why not using another strategy: just let ezjscore add its own optimizers, that will run all the time, but just set [eZJSCore] Packer = 1
That way, the optimizers of ezjscore will run yes, but will do nothing... (see the diff, some code has been added to handle the packLevel == 1 case)

As a result, the developer will be able to use his/her own custom extensions, whatever the order for loading those!

=> What do you think of this @andrerom ?

**Comment by jdespatis** [comment]
Here it is with this proposal to understand the idea

**Comment by jdespatis** [comment]
This pull can be considered as completed, with the 2 additional pull requests:
1/ https://github.com/ezsystems/ezpublish/pull/269
2/ https://github.com/ezsystems/ezpublish/pull/270

@bdunogier , @andrerom feel free to comment

**Comment by jdespatis** [comment]
Hum wait a minute, I encounter a problem by setting Packer=1:
when a user does something like {ezscript('djdojo::init',,,,0)} to force the pack level to 0 to add inline javascript code, then finally it gets overidden by 1 (because of Packer=1)

Strange... Investigation is needed...

**Comment by jdespatis** [comment]
Well here is the behavior: if ezjscore.ini[eZJSCore]\Packer is set to a value (1, 2 or 3), it overrides the value a user may specify in ezscript\* and ezcss\* template operators

It's due to the fact it's not possible to set a packLevel per file
Even if the user uses {ezscript()} template operator with only one file and set a specific pack_level, the latter will be overridden by ezjscore.ini[eZJSCore]\Packer 

So impossible to use the hint to set Packer=1 to force ezjscore to pack but not optimize

=> therefore solution 1 is to add a key in the Listeners is a working solution, as indicated by @andrerom , this way for example:

```
Listeners[css_optimizer]=optimize/css@ezjscCssOptimizer::optimize
```

Then a developer will be able to deactivate the optimizer from ezjscore by setting Listeners[css_optimizer]=
But the problem is that he/she can only do it once, and it can be messy if there're several css optimizer running (which is my case)

=> or solution 2 is to add a new setting in ezjscore.ini[eZJSCore]\Optimize set to 'enabled' or 'disabled' (set to 'enabled' by default)
and both js/css optimizers from ezjscore check this setting to know if they have to optimize or not

That way a developer, even if using several custom css optimizers will be able to fully deactivate optimization from ezjscore and control everything

From my pov, I think solution 2 is the best

or maybe a better solution, feel free to explain it...

@andrerom how to say... I need help ? :)

**Comment by andrerom** [comment]
In the first case you can do this if you need more then one:

```
Listeners[css_optimizer]=optimize/css@ezjscCssOptimizer::optimize
Listeners[css_optimizer2]=optimize/css@ezjscCssOptimizer::optimizeSecondSlowPass
```

The string can be whatever you want as long as it is unique, it's
purely for override use.

This is potentially a bit cleaner in the ini handling in the api where multi dimensional arrays are supported.

See: https://github.com/ezsystems/ezp-next/blob/master/ezp/Base/settings/service.ini#L108

Example of possible use:

```
Listeners[css_optimizer][]=<optimizer1>
Listeners[css_optimizer][]=<optimizer2>
```

**Comment by jdespatis** [comment]
Well my little mind has just understood the keyword logic, I've gotten wrong because of the generic keyword name: 'css_optimizer'.
Indeed a stupid developer (please don't look at me..) may understand to override it with:
Listeners[css_optimizer]=

_BUT_ the latter may also understand that he/she needs to feed again the Listeners[css_optimizer] array, which is completely stupid...
This code is just enough:

```
Listeners[css_optimizer]=
Listeners[]=optimize/css@callback
```

To keep clear this keyword is only for override use and is not intended to have a meaning, I've re-introduced this logic but with a very specific keyword to ezjscore like:

```
Listeners[ezjscore_css_optimizer]=...
Listeners[ezjscore_javascript_optimizer]=...
```

And also a small other pull request to avoid a warning, as a listener may be empty with this override process:
https://github.com/ezsystems/ezpublish/pull/271

Well now, all pull requests can be reviewed @andrerom , it runs nicely on my dev laptop

in the future, when multiarray will be supported, this logic could evolve

**Comment by jdespatis** [comment]
@andrerom Fell free to say if something is missing for this pull request to be merged, so that I can add it... :)

**Comment by andrerom** [comment]
For me it is just the fact that it is a breaking change, it removes a setting in exchange for another.
But on the upside it is possible to setup your extension to work with both the new and old setting though (I think).

**Comment by jdespatis** [comment]
Yes it breaks things, but event logic is the future, let's move to it !

But you're right, we can adopt a more smooth way to go: let the 2 logic work for some moment, and then deactivate completely the old logic
=> I'll change the pull, in order to keep the old logic work (for the transition)

**Comment by jdespatis** [comment]
@andrerom here we are, with last commit, the 2 logic are working, so the pull request doesn't break things at all
the 2 logics could work for 4.7, and CssOptimizer, and JavascriptOptimizer (old logic), could be removed in 4.8 (for example)

The side effect of last commit, is that css and js are optimized 2 times each !, which is not a problem as js/css optimizer in ezjscore is a safe process

Tell me, if you want me to comment the settings in site.ini to disable the Events (that would be activated in 4.8 for example), and then only use the old logic (a custom extension can easily activate back the new logic for ezjscore)

**Comment by andrerom** [comment]
> Yes it breaks things, but event logic is the future, let's move to it !

ehm.. just because it uses ezpEvent does not mean it will work on 5.x.

The last commit made it worse however, now it is not possible to define your settings that works for both 4.6 and 4.7 w/o keep using a deprecated setting :)

So lets step back a bit, what value does this actually bring to the solution other then just using ezpEvent instead?

**Comment by andrerom** [comment]
Closing as this PR seems to be dead and we are about to merge "system-extensions" like this one into eZ Publish 4.x repository.

**Pull request closed without merging**